cloud/services/virtualmachines/virtualmachines.go

```diff
@@ -158,10 +168,50 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 		return errors.Wrapf(err, "cannot create vm")
 	}
 
+	if vmSpec.Identity == infrav1.VMIdentitySystemAssigned {
+		err = s.createRoleAssignmentForIdentity(ctx, vmSpec.Name)
+		if err != nil {
+			return errors.Wrapf(err, "cannot create vm")
+		}
+	}
+
 	klog.V(2).Infof("successfully created VM %s ", vmSpec.Name)
 	return nil
 }
 
+func (s *Service) createRoleAssignmentForIdentity(ctx context.Context, vmName string) error {
+	resultVM, err := s.Client.Get(ctx, s.Scope.ResourceGroup(), vmName)
+	if err != nil {
+		return errors.Wrapf(err, "cannot get vm to assign role to system assigned identity")
+	}
+
+	// here should take the identity and create a role for it
+	roleClient := newRoleClient(s.Scope.SubscriptionID, s.Scope.Authorizer)
+	scope := fmt.Sprintf("/subscriptions/%s/", s.Scope.SubscriptionID)
+	// Azure built-in roles https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles
+	contributorRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c", s.Scope.SubscriptionID)
+	params := authorization.RoleAssignmentCreateParameters{
+		Properties: &authorization.RoleAssignmentProperties{
+			RoleDefinitionID: to.StringPtr(contributorRoleDefinitionID),
+			PrincipalID:      to.StringPtr(*resultVM.Identity.PrincipalID),
+		},
+	}
+	_, err = roleClient.Create(ctx, scope, string(uuid.NewUUID()), params)
+	if err != nil {
+		return errors.Wrapf(err, "cannot assign role to vm system assigned identity")
+	}
+
+	klog.V(2).Infof("successfully created Role assignment for generated Identity for VM %s ", vmName)
+	return nil
+}
+
+func newRoleClient(subscriptionID string, authorizer autorest.Authorizer) authorization.RoleAssignmentsClient {
```

Comment:
can this be moved into a client.go file in its own services/ package so that 1) it can be reused by other services in the future if needed, and mostly 2) we don't have dependencies on the Azure SDK for go in the virtualmachines.go file as we are purposefully breaking those appart for https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/416

You can then add the client in virtualmachines/service.go to use it in the function above. See how we use `PublicIPsClient` below as an example.

Reply:
I started doing that, but then decided not to because nothing else would call it, but yeah I can make that change